### PR TITLE
feat(units): delete child records on unit or app removal

### PR DESCRIPTION
**Pull request by wallyworld**
When an app or unit is deleted, rows with FK to the app or unit need to be removed.
For most things, we just start out by deleting the affected rows directly.
This can be adjusted as the implementation progresses. We have the DDL in place but not the data yet.

For secrets, the deletion is more complicated so we defer to the service.
We also need to cleanup external secret content. We use the previously added features of the secret service. Before deleting an app or unit, we gather the affected secret uris and delete them. We also gather a slice of clean up funcs to run after the db deletions to remove the external content and backend references.

Note: a fix was made to the cleanup func context in the secrets domain. There's also an issue with calling the CleanupSecrets method on the backend provider which has had a work around added and a TODO left.

A change was needed for unit deletion - `RemoveUnitMaybeApplication` is not longer required which fits better with the mongo workflow for now. The deletion stuff is still in progress so a few TODOs are added.

Where possible, the new internal errors package is used. However, we retain the use of the original `juju/errors` where there's already extensive use - cut over will be done holistically.

The end result is that the recently added bash ci tests for secret deletion in 3.6 now pass in main.

## QA steps

`./main.sh -v -p k8s secrets_k8s`

## Links

**Jira card:** JUJU-

### CODE REVIEW

File: domain/application/service/application.go [hunk 1]
```diff
@@ -323,8 +327,56 @@ func (s *ApplicationService) AddUnits(ctx context.Context, name string, units ..
...
+	// Get unit owned secrets.
+	uris, err := s.st.GetSecretsForOwners(ctx, application.NilApplicationOwners, application.UnitSecretOwners{unitName})
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting unit owned secrets for %q", unitName)
+	}
+	// Delete unit owned secrets.
+	var cleanups []func(context.Context)
+	for _, uri := range uris {
+		s.logger.Debugf("deleting unit %q secret: %s", unitName, uri.ID)
+		cleanup, err := s.secretService.InternalDeleteSecret(ctx, uri, secretservice.DeleteSecretParams{
```
Comment (ycliuhw) [inline domain/application/service/application.go:355]: Given we already have secretService here, should we implement  GetSecretsForOwners in secret service?

File: domain/application/service/application.go [hunk 2]
```diff
@@ -480,8 +534,51 @@ func (s *ApplicationService) insertCAASUnit(
...
+	var cleanups []func(context.Context)
+	for _, uri := range uris {
+		s.logger.Debugf("deleting application %q secret: %s", name, uri.ID)
+		cleanup, err := s.secretService.InternalDeleteSecret(ctx, uri, secretservice.DeleteSecretParams{
+			// TODO(units) - access is expected to be a unit
+			// It is passed to CleanupSecrets() on k8s and is used
+			// to name the service account..
+			Accessor: secretservice.SecretAccessor{
+				Kind: secretservice.UnitAccessor,
+				ID:   name + "/0",
```
Comment (ycliuhw) [inline domain/application/service/application.go:568]: Should it be the leader unit?

Comment (wallyworld) [reply domain/application/service/application.go:568]: no, in 3.6 it is the unit which makes the api call. it's only used to name the name for the k8s service account.
but it currently must be a unit.
this is a temp fix until we can decide the best way to adapt it to 4.0

File: domain/application/service/application.go (hunk 1 above)
Comment (wallyworld) [reply domain/application/service/application.go:358]: yeah, ok. moved the logic across to the secret service

**Review by ycliuhw — APPROVED** [review]
LGTM, ty!

File: domain/application/state/application.go [hunk 3]
```diff
@@ -274,13 +273,8 @@ WHERE name = $applicationName.name
...
 // If the application still has units, as error satisfying [applicationerrors.ApplicationHasUnits]
 // is returned.
-func (st *ApplicationState) DeleteApplication(ctx context.Context, name string) error {
-	db, err := st.DB()
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+func (st *ApplicationState) DeleteApplication(ctx domain.AtomicContext, name string) error {
```
Comment (manadart) [inline domain/application/state/application.go:276]: This is an example of incorrect use of a method with an `AtomicContext`. 

Just describing it in plain English shows that it has logic that should be in the service layer.
- Get the UUID for the application name.
- Check that no units exist for it.
- Delete all references.
- Delete the record.
